File: packages/engine/src/pots.ts
export interface PotLayer {
  /** Chips in this pot layer. */
  amount: number;
  /** Seat indices eligible to win this pot (did not fold and contributed to this layer). */
  eligible: number[];
}

export interface Contribution {
  seat: number;
  /** Total chips put into the pot this hand (committed). */
  amount: number;
  folded: boolean;
}

/**
 * Build main + side pots from per-seat contributions.
 * Folded players still contribute chips but are not eligible to win.
 */
export function buildSidePots(contributions: Contribution[]): PotLayer[] {
  const positive = contributions.filter((c) => c.amount > 0);
  if (positive.length === 0) return [];

  const levels = [...new Set(positive.map((c) => c.amount))].sort((a, b) => a - b);
  const pots: PotLayer[] = [];
  let prev = 0;

  for (const level of levels) {
    const contributors = positive.filter((c) => c.amount >= level);
    const eligible = contributors.filter((c) => !c.folded).map((c) => c.seat);
    const amount = (level - prev) * contributors.length;
    if (amount > 0) {
      pots.push({ amount, eligible });
    }
    prev = level;
  }

  // Merge adjacent pots with identical eligibility for cleaner display
  const merged: PotLayer[] = [];
  for (const pot of pots) {
    const last = merged[merged.length - 1];
    if (
      last &&
      last.eligible.length === pot.eligible.length &&
      last.eligible.every((s, i) => s === pot.eligible[i])
    ) {
      last.amount += pot.amount;
    } else {
      merged.push({ amount: pot.amount, eligible: [...pot.eligible] });
    }
  }
  return merged;
}

export interface PotAward {
  seat: number;
  amount: number;
  /** Best-hand category name at showdown, or "Uncontested" when everyone else folded. */
  handName?: string;
}

/**
 * Award pots given a comparable hand rank per seat (higher wins).
 * Odd chips go to the earliest eligible seat clockwise from dealerButton+1.
 */
export function awardPots(
  pots: PotLayer[],
  ranks: Map<number, number>,
  dealerButton: number,
  seatCount: number,
): PotAward[] {
  const totals = new Map<number, number>();

  for (const pot of pots) {
    if (pot.amount <= 0 || pot.eligible.length === 0) continue;

    let best = -1;
    const winners: number[] = [];
    for (const seat of pot.eligible) {
      const r = ranks.get(seat) ?? -1;
      if (r > best) {
        best = r;
        winners.length = 0;
        winners.push(seat);
      } else if (r === best) {
        winners.push(seat);
      }
    }
    if (winners.length === 0) continue;

    const share = Math.floor(pot.amount / winners.length);
    let remainder = pot.amount - share * winners.length;

    // Order winners clockwise from left of dealer for odd chips
    const ordered = [...winners].sort((a, b) => {
      const da = (a - dealerButton - 1 + seatCount * 2) % seatCount;
      const db = (b - dealerButton - 1 + seatCount * 2) % seatCount;
      return da - db;
    });

    for (const seat of ordered) {
      let amt = share;
      if (remainder > 0) {
        amt += 1;
        remainder -= 1;
      }
      totals.set(seat, (totals.get(seat) ?? 0) + amt);
    }
  }

  return [...totals.entries()].map(([seat, amount]) => ({ seat, amount }));
}

File: apps/android/core/engine/src/main/java/com/pokr/android/engine/Pots.kt
package com.pokr.android.engine

data class PotLayer(
    /** Chips in this pot layer. */
    val amount: Int,
    /** Seat indices eligible to win this pot (did not fold and contributed to this layer). */
    val eligible: List<Int>,
)

data class Contribution(
    val seat: Int,
    /** Total chips put into the pot this hand (committed). */
    val amount: Int,
    val folded: Boolean,
)

/**
 * Build main + side pots from per-seat contributions.
 * Folded players still contribute chips but are not eligible to win.
 */
fun buildSidePots(contributions: List<Contribution>): List<PotLayer> {
    val positive = contributions.filter { it.amount > 0 }
    if (positive.isEmpty()) return emptyList()

    val levels = positive.map { it.amount }.distinct().sorted()
    val pots = mutableListOf<PotLayer>()
    var prev = 0

    for (level in levels) {
        val contributors = positive.filter { it.amount >= level }
        val eligible = contributors.filter { !it.folded }.map { it.seat }
        val amount = (level - prev) * contributors.size
        if (amount > 0) {
            pots.add(PotLayer(amount, eligible))
        }
        prev = level
    }

    // Merge adjacent pots with identical eligibility for cleaner display
    val merged = mutableListOf<PotLayer>()
    for (pot in pots) {
        val last = merged.lastOrNull()
        if (last != null &&
            last.eligible.size == pot.eligible.size &&
            last.eligible == pot.eligible
        ) {
            merged[merged.lastIndex] = last.copy(amount = last.amount + pot.amount)
        } else {
            merged.add(pot.copy(eligible = pot.eligible.toList()))
        }
    }
    return merged
}

data class PotAward(
    val seat: Int,
    val amount: Int,
    /** Best-hand category name at showdown. Omitted when the pot is won uncontested. */
    val handName: String? = null,
)

/**
 * Award pots given a comparable hand rank per seat (higher wins).
 * Odd chips go to the earliest eligible seat clockwise from dealerButton+1.
 */
fun awardPots(
    pots: List<PotLayer>,
    ranks: Map<Int, HandRank>,
    dealerButton: Int,
    seatCount: Int,
): List<PotAward> {
    val totals = mutableMapOf<Int, Int>()

    for (pot in pots) {
        if (pot.amount <= 0 || pot.eligible.isEmpty()) continue

        var best = -1
        val winners = mutableListOf<Int>()
        for (seat in pot.eligible) {
            val r = ranks[seat] ?: -1
            if (r > best) {
                best = r
                winners.clear()
                winners.add(seat)
            } else if (r == best) {
                winners.add(seat)
            }
        }
        if (winners.isEmpty()) continue

        val share = pot.amount / winners.size
        var remainder = pot.amount - share * winners.size

        // Order winners clockwise from left of dealer for odd chips
        val ordered = winners.sortedBy { seat ->
            (seat - dealerButton - 1 + seatCount * 2) % seatCount
        }

        for (seat in ordered) {
            var amt = share
            if (remainder > 0) {
                amt += 1
                remainder -= 1
            }
            totals[seat] = (totals[seat] ?: 0) + amt
        }
    }

    return totals.entries.map { (seat, amount) -> PotAward(seat, amount) }
}
